lld/ELF/NanoMipsTransformations.cpp

```diff
@@ -0,0 +1,1441 @@
+      sec->nanoMipsRelaxAux->freeBytes = 0;
+
+      auto *abiFlagsSec = NanoMipsAbiFlagsSection<ELFT>::get();
+
+      auto *obj = sec->getFile<ELFT>();
+
+      sec->nanoMipsRelaxAux->pcrel = obj ? isNanoMipsPcRel<ELFT>(obj) : true;
+
+      sec->nanoMipsRelaxAux->fullNanoMipsISA = false;
+      if (!abiFlagsSec)
+        error("Abi flags section not created, it is needed to determine "
+              "whether full nanoMIPS ISA is used!");
+      else
+        sec->nanoMipsRelaxAux->fullNanoMipsISA =
+            abiFlagsSec->isFullNanoMipsISA(sec);
+    }
+  }
+
+  DenseSet<Defined *> seenWrapped;
+  for (InputFile *file : ctx.objectFiles) {
+    for (Symbol *sym : file->getSymbols()) {
+      auto *d = dyn_cast<Defined>(sym);
+      if (!d || d->file != file)
+        continue;
+      if (auto *sec = dyn_cast_or_null<InputSection>(d->section)) {
+        if ((sec->flags & (SHF_EXECINSTR | SHF_ALLOC)) !=
+                (SHF_EXECINSTR | SHF_ALLOC) ||
+            !(sec->type & SHT_PROGBITS) || !sec->nanoMipsRelaxAux)
+          continue;
+        // Note: Wrapped symbols change their real symbol, and we
+        // get two pointers to the same wrap symbol, which we don't
+        // need in the Aux struct as we are going to update that
+        // symbol twice then.
+        if (d->getName().starts_with("__wrap_"))
+          if (!seenWrapped.insert(d).second)
+            continue;
+        sec->nanoMipsRelaxAux->anchors.push_back({d, false});
+        if (d->size > 0)
+          sec->nanoMipsRelaxAux->anchors.push_back({d, true});
+      }
+    }
+  }
+}
+
+template <class ELFT>
+void NanoMipsTransformController<ELFT>::scanAndTransform(
+    InputSection *sec) const {
+
+  LLVM_DEBUG(llvm::dbgs() << "\nStarted transforming " << sec->name
+                          << " section\n\n");
+  bool seenNoRelax = false;
+
+  const uint32_t bits = config->wordsize * 8;
+  uint64_t secAddr = sec->getOutputSection()->addr + sec->outSecOff;
+  // Need to traverse relocations this way because at transform we may
+  // invalidate the iterator and add some relocs
+  for (uint32_t relNum = 0; relNum < sec->relocs().size(); relNum++) {
+    Relocation &reloc = sec->relocs()[relNum];
+
+    // TODO: Check out if relocation with the same offset as R_NANOMIPS_NORELAX
+    // should or shouldn't be relaxed, NORELAX and RELAX are put after the
+    // relocation referring to the instruction by the gnu assembler, as are
+    // FIXED, INSN32, INSN16...
+    if (reloc.type == R_NANOMIPS_NORELAX) {
+      seenNoRelax = true;
+      continue;
+    }
+
+    if (reloc.type == R_NANOMIPS_RELAX) {
+      seenNoRelax = false;
+      continue;
+    }
+
+    if (seenNoRelax)
+      continue;
+
+    if (reloc.type == R_NANOMIPS_ALIGN) {
+      this->align(sec, reloc, relNum);
+      continue;
+    }
+
+    const NanoMipsRelocProperty *relocProp =
+        relocPropertyTable.getRelocProperty(reloc.type);
+    if (!relocProp)
```

Comment:
There isn't. They are just left as they are (instructions referring to it remain unchanged). This could be a problem if  this instruction needs to be expanded, so this will trigger an error later (which is not good).

So either this is a good place (or at the beginning of the for loop) for an assert, or even at the end when actually relocating, if R_NANOMIPS_GPREL16_S2 is not expected at all. Thanks!